src/passes/J2CLItableMerging.cpp

```diff
@@ -104,20 +104,33 @@ struct J2CLItableMerging : public Pass {
         continue;
       }
 
-      auto type = heapType.getStruct();
-      if (typeNameInfo.fieldNames.empty() ||
-          !typeNameInfo.fieldNames[0].equals("vtable")) {
-        continue;
-      }
-      if (typeNameInfo.fieldNames.size() < 1 ||
-          !typeNameInfo.fieldNames[1].equals("itable")) {
+      if (typeNameInfo.fieldNames.empty()) {
         continue;
       }
 
-      auto vtabletype = type.fields[0].type.getHeapType();
-      auto itabletype = type.fields[1].type.getHeapType();
+      // The vtable may either be the first field or the custom descriptor.
+      std::optional<HeapType> vtabletype;
+      std::optional<HeapType> itabletype;
+      auto type = heapType.getStruct();
+      if (auto descriptor = heapType.getDescriptorType()) {
+        if (!typeNameInfo.fieldNames[0].equals("itable")) {
+          continue;
+        }
+        vtabletype = descriptor;
+        itabletype = type.fields[0].type.getHeapType();
+      } else {
+        if (!typeNameInfo.fieldNames[0].equals("vtable")) {
+          continue;
+        }
+        if (!typeNameInfo.fieldNames.count(1) ||
+            !typeNameInfo.fieldNames[1].equals("itable")) {
```

Comment:
It seems we should have a similar check before looking up the name of field 0. Also, we can avoid double lookup by doing this:
```suggestion
        if (auto it = typeNameInfo.fieldNames.find(1);
            it == typeNameInfo.fieldNames.end() || it->second.equals("itable")) {